lib/bloc/account/add_funds_bloc.dart

```diff
@@ -11,19 +14,65 @@ class AddFundsBloc {
   final _addFundResponseController = new StreamController<AddFundResponse>();
   Stream<AddFundResponse> get addFundResponseStream => _addFundResponseController.stream;
 
+  final _moonPayOrderController = new StreamController<MoonpayOrder>();
+  Sink<MoonpayOrder> get moonPayOrderSink => _moonPayOrderController.sink;
+  Stream<MoonpayOrder> get moonPayOrderStream => _moonPayOrderController.stream;
+
+  final _orderController = new BehaviorSubject<MoonpayOrder>();
+  Sink<MoonpayOrder> get orderSink => _orderController.sink;
+
   AddFundsBloc(String userID){
-    BreezBridge breezLib = new ServiceInjector().breezBridge;
+    ServiceInjector injector = ServiceInjector();
+    BreezBridge breezLib = injector.breezBridge;
     _addFundRequestController.stream.listen((request){
       _addFundResponseController.add(null);
         breezLib.addFundsInit(userID)
           .then((reply) => _addFundResponseController.add(new AddFundResponse(reply)))
           .catchError(_addFundResponseController.addError);
       })
       .onDone(_dispose);
+
+    _initializeWithMoonpayOrder(injector);
+    _listenToMoonpayOrder(injector);
+  }
+
+  _initializeWithMoonpayOrder(ServiceInjector injector) {
+    injector.sharedPreferences.then((preferences) {
+      DateTime orderDate = DateTime.fromMillisecondsSinceEpoch(preferences.getInt("pendingMoonpayOrderTimestamp") ?? 0);
+      // Expire order that lasted longer than 15 minutes
+      if (DateTime.now().difference(orderDate).inMinutes >= 15) {
+        _removeMoonpayOrder(preferences);
+        _moonPayOrderController.add(MoonpayOrder(null, null));
+      } else {
+        _moonPayOrderController.add(MoonpayOrder(
+            preferences.getString("pendingMoonpayOrderAddress") ?? null, preferences.getInt("pendingMoonpayOrderTimestamp") ?? null));
+      }
+    });
+  }
+
+  _removeMoonpayOrder(SharedPreferences preferences) {
+    preferences.remove("pendingMoonpayOrderAddress");
+    preferences.remove("pendingMoonpayOrderTimestamp");
+  }
+
+  _listenToMoonpayOrder(ServiceInjector injector) {
+    _orderController.stream.listen((order) async {
+      injector.sharedPreferences.then((preferences) {
+        _saveMoonpayOrder(preferences, order);
+        _moonPayOrderController.add(order);
+      });
+    });
+  }
+
+  _saveMoonpayOrder(SharedPreferences preferences, MoonpayOrder order) {
+    preferences.setString("pendingMoonpayOrderAddress", order.address);
```

Comment:
I think it is better that MoonpayOrder class will handle its own json serialization and we can write its value under one key.

Reply:
[Fixed](https://github.com/breez/breezmobile/pull/162/commits/8f9ff8cb02826657a80bf08ddb67b2dd7453a0f4)